File: leetcode-75/1768_merge_string_alternately/merge_string_alternately.cpp
#include <string>
#include <iostream>
using namespace std;

class Solution{
  public:
    string mergeAlternately(string word1, string word2) {
      // reserve the exact capacity to avoid multiple alloctions
      string result;
      result.reserve(word1.size() + word2.size());

      int i = 0, j = 0;
     // merge characters alternately 
      while (i < word1.size() && j < word2.size()) {
        result.push_back(word1[i++]);
        result.push_back(word2[j++]);
      }
      // append the remaining chars from word1, if any
      while (i < word1.size()) {
        result.push_back(word1[i++]);
      }
      // append the remaining chars from word2, if any
      while (j < word2.size()) {
        result.push_back(word2[j++]);
      }
      return result;
    }
};


int main() {
    string word1 = "abc";
    string word2 = "xyz";
    Solution sol;
    string mergedString = sol.mergeAlternately(word1, word2);

    cout << "Merged String: " << mergedString << endl;

    return 0;
}

File: leetcode-75/1768_merge_string_alternately/merge_string_alternately.rs
pub fn merge_alternately(word1: String word2: String) -> String {
    let mut result = String::with_capacity(word1.len() + word2.len());
    let mut i = 0;
    let mut j = 0;
    let word1_chars: Vec<char> = word1.chars().collect();
    let word2_chars: Vec<char> = word2.chars().collect();


    while i < word1_chars.len() && j < word2_chars.len() {
        result.push(word1_chars[i]);
        result.push(word2_chars[i]);
        i += 1;
        j += 1;
    }
    result.extend(word1_chars[i..].iter());
    result.extend(word2_chars[i..].iter());
    
    result
}
